### DISCUSSION TITLE
Accuracy: Google Coral vs TensorRT models.

### DISCUSSION THREAD

**Discussion by twister36** [blakeblackshear/frigate#9074] [Ask A Question]
Anyone else running two instances of Frigate for comparison?

I have one instance that is running on a Google Coral and another running on a TensorRT (Nvidia GPU). I have found that the Google Coral is x10 better at properly identifying objects with less false positives.

For example, the TensorRT models falsely identify my dog as a cat, but the same video feed the Google Coral will see her as a dog. The TensorRT randomly picks up objects and identifies them up as "people", even though they are not. 

For TensorRT, I've switch models from YOLOV7-320 and tried YOLOV7-640 and YOLOV7x-640

Environment 1:
- Raspberry Pi 4 (8GB)
- Google Corel (USB)
- Frigate 12 (0.12.1-367D724)

Environment 2:
- Windows 11 Pro 
- NVIDIA GeForce RTX 4060 Ti
- Frigate 13 - Beta 7 (0.13.0-C35C7DA)

Thoughts?

**Comment by NickM-27** [comment]
In general the models are not trained on camera examples so results will greatly vary from user to user. It is interesting to hear your results as well as others

**Reply by twister36** [reply]
@NickM-27  - Do you suggest that I try all the other TensorRT models and see which one performs the best then?

In the end I plan to fall back to my Google Coral, as it seems to be the best for properly identifying. 

Also the Nvidia GPU appears to be out performing the Google Coral:

Google Coral:
- 4 cameras (704x480 resolution)
- Inference Speed: 17ms

TensorRT - NVIDIA GeForce RTX 4060 Ti:
- 9 cameras (704x480 resolution)
- Inference Speed: 3-6ms

**Comment by bairhys** [comment]
It is interesting how your RPi4 USB Coral is 17ms, At the moment I have a M.2 Coral (standard Frigate Coral setup) with a i5-7500 and averages 8ms:

![image](https://github.com/blakeblackshear/frigate/assets/33058425/1407e5d0-61d5-4dad-876c-f942d22981cf)

Different models on the same hardware change the inference speed, The built in Frigate OpenVino SSDLite MobileNet model I get average 12ms:

![image](https://github.com/blakeblackshear/frigate/assets/33058425/d3da4473-5e2f-44d5-8fbb-b7ba16280cd2)

when using a OpenVino yolox-tiny model I was getting average 41ms:

![image](https://github.com/blakeblackshear/frigate/assets/33058425/7b4f4b89-a8fc-4746-ba27-f49c62cbb8b1)

But from memory I didn't see a obvious improvement to accuracy using yolox-tiny over SSDLite MobileNet

**Reply by twister36** [reply]
@bairhys - Are you running your Frigate in HA? Or is it a dedicated instance? For mine, the Raspberry Pi 4 is a dedicated box running Frigate, it does nothing else. For the Nvidia GPU machine, this is Windows machine, and does very well, as long as I don't need the GPU for anything else (i.e. Photoshop, etc). 

Since this post, I've installed Frigate on a 3rd system, an OptiPlex i7-8700 with Google Coral and it too is around 8ms.

**Reply by bairhys** [reply]
My Frigate container runs on a i5-7500 with Ubuntu server and I have HA on a Pi4

**Reply by twister36** [reply]
Are you using hardware acceleration for your intel CPU?

ffmpeg:
  hwaccel_args: preset-vaapi

**Reply by bairhys** [reply]
Yeah I use `hwaccel_args: preset-vaapi` to use intel gpu

**Comment by jmtatsch** [comment]
@twister36 Try to set the input size to the size the model has been trained for:

model:
  path: /config/model_cache/tensorrt/yolov7-320.trt
  input_tensor: nchw
  input_pixel_format: rgb
  width: 320
  height: 320

not 704x480

For me that nvidia classifier is more accurate, but burns way more power.

**Reply by twister36** [reply]
@jmtatsch - Thanks for the reply. Yes, I always match them... if I'm using YOLO 640, then my width/height is set to 640 as well.

**Comment by The-NinthWave** [comment]
@NickM-27 

Nice to see someone else's results because I installed a Coral this week and was running a Quadro P600 for a year and I think it not good :

Environment1
QuadroP600, yolov4-tiny-416, inference 10ms
```
detectors:
  tensorrt:
    type: tensorrt
model:
  path: /trt-models/yolov4-tiny-416.trt
  input_tensor: nchw
  input_pixel_format: rgb
  width: 416
  height: 416
```

Environment2
Cora edge-tpu,default, inference 6,5ms
```
detectors:
  coral:
    type: edgetpu
    device: pci
```
We parked the car in the garage an hour a go and it was not even detected.

Anything I can do to improve efficiency of the Coral ?

**Reply by NickM-27** [reply]
The accuracy has to do with the model that is run, not the hardware that runs the model

**Comment by incith** [comment]
Is it, or is there a list, and is it worthwhile to generate a model that is the same as the sub stream resolutions? E.g. is using the yolov7-640 on a 640x480 sub stream advisable, or should I be looking for a different model?

**Comment by ors86** [comment]
So witch is the winner for odject detection betweenGoogle Coral vs TensorRT (nvidia GPU) ?

**Reply by NickM-27** [reply]
There is no objective absolute winner. The results are subjective and vary from person to person, camera to camera, etc.
